stan/math/rev/core/var.hpp

```diff
@@ -177,12 +149,284 @@ class var_value {
+   * instances.
+   */
+  vari_type* vi_;
+
+  /**
+   * Return `true` if this variable has been
+   * declared, but not been defined.  Any attempt to use an
+   * undefined variable's value or adjoint will result in a
+   * segmentation fault.
+   *
+   * @return <code>true</code> if this variable does not yet have
+   * a defined variable.
+   */
+  inline bool is_uninitialized() { return (vi_ == nullptr); }
+
+  /**
+   * Construct a variable for later assignment.
+   *
+   * This is implemented as a no-op, leaving the underlying implementation
+   * dangling.  Before an assignment, the behavior is thus undefined just
+   * as for a basic double.
+   */
+  var_value() : vi_(nullptr) {}
+
+  /**
+   * Construct a variable from the specified floating point argument
+   * by constructing a new `vari_value<value_type>`. This constructor is only
+   * valid when `S` is convertible to this `vari_value`'s `value_type`.
+   * @tparam S A type that is convertible to `value_type`.
+   * @param x Value of the variable.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  var_value(S&& x) : vi_(new vari_type(std::forward<S>(x), false)) {}  // NOLINT
+
+  /**
+   * Copy constructor for var_val.
+   * @tparam S type of the value in the `var_value` to assing
+   * @param other the value to assign
+   * @return this
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr,
+            require_all_plain_type_t<T, S>* = nullptr>
+  var_value(const var_value<S>& other) : vi_(other.vi_) {}
+
+  /**
+   * Construct a `var_value` with a plain type
+   *  from another `var_value` containing an expression.
+   * @tparam S type of the value in the `var_value` to assing
+   * @param other the value to assign
+   * @return this
+   */
+  template <typename S, typename T_ = T,
+            require_convertible_t<S&, value_type>* = nullptr,
+            require_not_plain_type_t<S>* = nullptr,
+            require_plain_type_t<T_>* = nullptr>
+  var_value(const var_value<S>& other) : vi_(new vari_type(other.vi_->val_)) {
+    reverse_pass_callback(
+        [this_vi = this->vi_, other_vi = other.vi_]() mutable {
+          other_vi->adj_ += this_vi->adj_;
+        });
   }
 
+  /**
+   * Construct a variable from a pointer to a variable implementation.
+   * @param vi A vari_value pointer.
+   */
+  var_value(vari_type* vi) : vi_(vi) {}  // NOLINT
+
+  /**
+   * Return a constant reference to the value of this variable.
+   *
+   * @return The value of this variable.
+   */
+  inline const auto& val() const { return vi_->val_; }
+
+  /**
+   * Return a reference to the derivative of the root expression with
+   * respect to this expression.  This method only works
+   * after one of the `grad()` methods has been
+   * called.
+   *
+   * @return Adjoint for this variable.
+   */
+  inline auto& adj() { return vi_->adj_; }
+  inline auto& adj() const { return vi_->adj_; }
+  inline auto& adj_op() { return vi_->adj_; }
```

Comment:
to avoid an unneeded eval in the var matrix multiply I has to add [`adj_op`](https://github.com/stan-dev/math/blob/develop/stan/math/prim/eigen_plugins.h#L167) to the eigen plugins to return back a non-const `CwiseUnaryOp`. This just is so that `var<matrix>` and `matrix<var>` can use the same code

Reply:
See discussion below

https://github.com/stan-dev/math/pull/2091#discussion_r493966109

Reply:
Hmm, would not making .adj() non-const solve the problem?

Reply:
Tagging @andrjohns as I think he has a better view on this. My answer is idt so

Reply:
@andrjohns Did you miss this? I would really prefere if we could have only 1 adj function.

Reply:
@andrjohns Did you miss this? I would really prefere if we could have only one adj function.

Reply:
I think he recently talked about it over here

https://github.com/stan-dev/math/pull/2106#discussion_r497631555